### PR TITLE
out_s3: add content options

### PR DESCRIPTION
**Pull request by zhonghui12**
Signed-off-by: Zhonghui Hu <zh0512xx@gmail.com>

Add content options for users to define `Content-Encoding` and `Content-Type`.

Part of #2700 
----
Enter `[N/A]` in the box, if an item is not applicable to your change.

**Testing**
Before we can approve your change; please submit the following in a comment:
- [ ] Example configuration file for the change
- [x] Debug log output from testing the change

- [ ] Attached [Valgrind](https://valgrind.org/docs/manual/quick-start.html) output that shows no leaks or memory corruption was found

**Documentation**

- [ ] Documentation required for this feature

----

Fluent Bit is licensed under Apache 2.0, by submitting this pull request I understand that this code will be released under the terms of that license.

### CODE REVIEW

**Comment by edsiper** [comment]
assigned to @PettitWesley

File: plugins/out_s3/s3.c [hunk 1]
```diff
@@ -101,6 +115,50 @@ struct flb_aws_header *create_canned_acl_header(char *canned_acl)
...
+    headers = flb_malloc(sizeof(struct flb_aws_header) * num_headers);
+    if (headers == NULL) {
+        flb_errno();
+        return NULL;
+    }
+
+    if (ctx->content_type != NULL) {
+        content_type_header.val = ctx->content_type;
+        content_type_header.val_len = strlen(ctx->content_type);
+        headers[n] = content_type_header;
```
Comment (PettitWesley) [inline plugins/out_s3/s3.c:136]: Since this plugin can make concurrent requests... I am pretty sure this is not safe. Multiple requests could use the header declared with `static` at the same time.

Your headers array is heap allocated memory. Instead of this:

```
 headers[n] = content_type_header;
```

Use memcpy to copy the static header into the array of memory and then modify it

File: plugins/out_s3/s3.c [hunk 2]
```diff
@@ -50,6 +51,19 @@ static struct multipart_upload *get_upload(struct flb_s3 *ctx,
 static struct multipart_upload *create_upload(struct flb_s3 *ctx,
                                               const char *tag, int tag_len);
 
+static struct flb_aws_header content_encoding_header = {
+    .key = "Content-Encoding",
```
Comment (PettitWesley) [inline plugins/out_s3/s3.c:55]: More on my comment below about concurrency. This header is safe to use directly in concurrent requests because its read-only. You won't be changing any of its fields. 

The `content_type_header` below is not concurrent safe since you write to it

File: plugins/out_s3/s3.c [hunk 3]
```diff
@@ -101,6 +115,50 @@ struct flb_aws_header *create_canned_acl_header(char *canned_acl)
...
+        content_type_header.val_len = strlen(ctx->content_type);
+        headers[n] = content_type_header;
+        n++;
+    }
+    if (ctx->compression != NULL) {
+        if (strcmp(ctx->compression, "gzip") != 0) {
+            flb_plg_error(ctx->ins, 
+                          "'gzip' is the only support value of compression field for now.");
+            flb_free(headers);
+            return NULL;
```
Comment (PettitWesley) [inline plugins/out_s3/s3.c:144]: Don't put this sort of user-input validation in this function. You want to do user validation in the init function. Once you are in this function then the input should be already validated.

File: plugins/out_s3/s3.c [hunk 4]
```diff
@@ -980,28 +1046,51 @@ static int s3_put_object(struct flb_s3 *ctx, const char *tag, time_t create_time
...
-            }
-            c = s3_client->client_vtable->request(s3_client, FLB_HTTP_PUT,
-                                                  uri, body, body_size,
-                                                  canned_acl_header, 1);
-            flb_free(canned_acl_header);
+        c = s3_client->client_vtable->request(s3_client, FLB_HTTP_PUT,
+                                              uri, final_body, final_body_size,
+                                              headers, num_headers);
+        if (ctx->canned_acl != NULL) {
+            flb_free(&headers[num_headers-1]);
```
Comment (zhonghui12) [inline plugins/out_s3/s3.c:1091]: Actually I am thinking about if it is a good way to free canned_acl_header. But I haven't thought about a better and more elegant way yet..

File: plugins/out_s3/s3.c [hunk 5]
```diff
@@ -101,6 +115,44 @@ struct flb_aws_header *create_canned_acl_header(char *canned_acl)
...
+    }
+
+    if (ctx->content_type != NULL) {
+        memcpy(&headers[n], &content_type_header, sizeof(content_type_header));
+        headers[n].val = ctx->content_type;
+        headers[n].val_len = strlen(ctx->content_type);
+        n++;
+    }
+    if (ctx->compression != NULL) {
+        headers[n] = content_encoding_header;
```
Comment (PettitWesley) [inline plugins/out_s3/s3.c:140]: you need to use memcpy here too

Comment (PettitWesley) [reply plugins/out_s3/s3.c:140]: headers is a block of memory that can store some number of flb_aws_headers structs. Its actually just raw memory:

```
flb_malloc(sizeof(struct flb_aws_header) * num_headers);
```

sizeof(struct flb_aws_header) just returns the number of bytes needed for one struct, so there's nothing special about your block of memory

Comment (PettitWesley) [reply plugins/out_s3/s3.c:140]: its just memory. 

And you need to copy the structure into this memory. In C, the `=` operator can only be used to copy memory for primitive types as far as I know. For structures, you must use memory copying functions.

Comment (zhonghui12) [reply plugins/out_s3/s3.c:140]: I see. Thanks.

File: plugins/out_s3/s3.c [hunk 6]
```diff
@@ -101,6 +115,44 @@ struct flb_aws_header *create_canned_acl_header(char *canned_acl)
     return acl_header;
 };
 
+static struct flb_aws_header *create_headers(struct flb_s3 *ctx, int num_headers)
+{
```
Comment (PettitWesley) [inline plugins/out_s3/s3.c:119]: In any language, it's good to design each function to do a specific job. As much as possible, you want to make sure it only does a single job, and encapsulates everything needed for that job. And then give it the fewest number of arguments possible

In this case- I think this function can calculate `num_headers` itself using the `ctx`. That encapsulates all behavior related to creating headers in this function. 

Outside you need to use the num_headers of course, so this function needs to return that information.

In C there is a very common pattern for doing this:

```
static int create_headers(struct flb_s3 *ctx, struct flb_aws_header **headers, int num_headers)
```

Basically, to return multiple pieces of information, at multiple pointer arguments. The return value is 0 or -1 depending on whether the function was successful or encountered an error.

You then call it like this:
```
struct flb_aws_header *headers;
int num_headers;

ret = create_headers(ctx, &headers, &num_headers);
```

And then in the function, allocate the headers as you do now, and then "return" them at the end like this:

```
*headers = s3_headers;
*num_headers = header_len;
```

Something like that.

File: plugins/out_s3/s3.c [hunk 7]
```diff
@@ -101,6 +115,44 @@ struct flb_aws_header *create_canned_acl_header(char *canned_acl)
...
+        n++;
+    }
+    if (ctx->compression != NULL) {
+        headers[n] = content_encoding_header;
+        n++;
+    }
+    if (ctx->canned_acl != NULL) {
+        canned_acl_header = create_canned_acl_header(ctx->canned_acl);
+        if (canned_acl_header == NULL) {
+            flb_plg_error(ctx->ins, "Failed to create canned ACL header");
```
Comment (PettitWesley) [inline plugins/out_s3/s3.c:146]: now that you have this function for creating the headers, it's okay to have a separate `create_canned_acl_header` function, but you probably could just copy all of its code to this function. That way you only have a single allocation of memory for headers.

File: plugins/out_s3/s3.c [hunk 8]
```diff
@@ -101,6 +115,44 @@ struct flb_aws_header *create_canned_acl_header(char *canned_acl)
...
+            flb_free(headers);
+            return NULL;
+        }
+        headers[n] = *canned_acl_header;
```
Comment (PettitWesley) [inline plugins/out_s3/s3.c:150]: See comment, this should use memcpy too, and should probably be refactored to work the same way as the content_type header.

File: plugins/out_s3/s3.c (hunk 4 above)
Comment (PettitWesley) [reply plugins/out_s3/s3.c:1091]: I am confused, why not free the entire header array- with:

```
flb_free(headers);
```

The headers array is a single block of memory

Comment (zhonghui12) [reply plugins/out_s3/s3.c:1091]: I checked online and it says: "You will need to free what you have allocated."  I free the entire header array and the canned_acl_header is also allocated by create_acl_header. So I want to free that one as well..

Comment (PettitWesley) [reply plugins/out_s3/s3.c:1091]: yeah see my other comments, you should refactor so that there is only one allocation.

(Also even if you allocate the canned header separately, this will not free it. If that doesn't make sense, call me and we can draw out what the memory actually looks like (or try that yourself)).

Comment (zhonghui12) [reply plugins/out_s3/s3.c:1091]: Thanks. I would try by myself first to see if I can understand it.

File: plugins/out_s3/s3.c [hunk 9]
```diff
@@ -980,28 +1045,39 @@ static int s3_put_object(struct flb_s3 *ctx, const char *tag, time_t create_time
     flb_sds_destroy(s3_key);
     uri = tmp;
 
+    if (ctx->use_put_object == FLB_TRUE && ctx->compression != NULL) {
+        if (strcmp(ctx->compression, "gzip") != 0) {
+            flb_plg_error(ctx->ins, 
+                          "'gzip' is the only support value of compression field for now");
+            flb_sds_destroy(uri);
+            return -1;
+        }
```
Comment (PettitWesley) [inline plugins/out_s3/s3.c:1054]: User input validation should be done in the init function `cb_s3_init`, that way if the config is invalid it will fail on startup so that the issue gets immediate notification

Comment (PettitWesley) [reply plugins/out_s3/s3.c:1054]: You should have two validations:
- gzip is the only supported value
- use_put_object must be enabled

File: plugins/out_s3/s3.c [hunk 10]
```diff
@@ -980,28 +1045,39 @@ static int s3_put_object(struct flb_s3 *ctx, const char *tag, time_t create_time
...
-                flb_sds_destroy(uri);
-                flb_plg_error(ctx->ins, "Failed to create canned ACL header");
-                return -1;
-            }
-            c = s3_client->client_vtable->request(s3_client, FLB_HTTP_PUT,
-                                                  uri, body, body_size,
-                                                  canned_acl_header, 1);
-            flb_free(canned_acl_header);
+        ret = create_headers(ctx, &headers, &num_headers);
+        if (ret == -1 || (headers == NULL && num_headers != 0)) {
```
Comment (PettitWesley) [inline plugins/out_s3/s3.c:1072]: the check should just be `if (ret < 0)`

your `create_headers` function should return `-1` for all failure cases

File: plugins/out_s3/s3.c [hunk 11]
```diff
@@ -1447,6 +1523,16 @@ static struct flb_config_map config_map[] = {
...
     },
+    {
+     FLB_CONFIG_MAP_STR, "compression", NULL,
+     0, FLB_FALSE, 0,
+    "Compression type for S3 objects. 'gzip' is the only support value for now."
+    },
+    {
+     FLB_CONFIG_MAP_STR, "content_type", NULL,
+     0, FLB_FALSE, 0,
+    "A standard MIME type describing the format of the contents."
```
Comment (PettitWesley) [inline plugins/out_s3/s3.c:1534]: Add something like: this will be set as the Content-Type HTTP header

just to make it very clear

File: plugins/out_s3/s3.c (hunk 9 above)
Comment (zhonghui12) [reply plugins/out_s3/s3.c:1054]: Got it.

File: plugins/out_s3/s3.c [hunk 12]
```diff
@@ -431,6 +482,25 @@ static int cb_s3_init(struct flb_output_instance *ins,
     if (tmp) {
         ctx->canned_acl = (char *) tmp;
     }
+
+    tmp = flb_output_get_property("compression", ins);
+    if (tmp) {
+        if (strcmp((char *) tmp, "gzip") != 0) {
+            flb_plg_error(ctx->ins, 
+                          "'gzip' is the only support value of compression field for now");
```
Comment (PettitWesley) [inline plugins/out_s3/s3.c:490]: I prefer: "'gzip' is currently the only supported value for 'compression'"

File: plugins/out_s3/s3.c [hunk 13]
```diff
@@ -431,6 +482,25 @@ static int cb_s3_init(struct flb_output_instance *ins,
...
     }
+
+    tmp = flb_output_get_property("compression", ins);
+    if (tmp) {
+        if (strcmp((char *) tmp, "gzip") != 0) {
+            flb_plg_error(ctx->ins, 
+                          "'gzip' is currently the only supported value for 'compression'");
+            return -1;
+        } else if (ctx->use_put_object == FLB_FALSE) {
+            flb_plg_error(ctx->ins, "use_put_object should be enabled");
```
Comment (PettitWesley) [inline plugins/out_s3/s3.c:493]: This does not give enough context: how about: "use_put_object must be enabled when compression is enabled"

File: plugins/out_s3/s3.c [hunk 14]
```diff
@@ -930,13 +1000,17 @@ static int s3_put_object(struct flb_s3 *ctx, const char *tag, time_t create_time
...
+    struct flb_aws_header *headers = NULL;
     char *random_alphanumeric;
     int append_random = FLB_FALSE;
     int len;
+    int ret;
+    int num_headers = 0;
     char *final_key;
     flb_sds_t uri;
     flb_sds_t tmp;
+    void *final_body;
```
Comment (PettitWesley) [inline plugins/out_s3/s3.c:1012]: why void pointer? it should be a `char *` like `body` I think

File: plugins/out_s3/s3.c [hunk 15]
```diff
@@ -1447,6 +1526,18 @@ static struct flb_config_map config_map[] = {
      0, FLB_FALSE, 0,
     "Predefined Canned ACL policy for S3 objects."
     },
+    {
+     FLB_CONFIG_MAP_STR, "compression", NULL,
+     0, FLB_FALSE, 0,
+    "Compression type for S3 objects. 'gzip' is the only support value for now. "
```
Comment (PettitWesley) [inline plugins/out_s3/s3.c:1532]: I don't like "for now" it doesn't sound like official language to me, sounds too conversational

"'gzip' is currently the only supported value"

File: plugins/out_s3/s3.c [hunk 16]
```diff
@@ -1447,6 +1526,18 @@ static struct flb_config_map config_map[] = {
...
+    "This will be set as the Content-Encoding HTTP header."
+    },
+    {
+     FLB_CONFIG_MAP_STR, "content_type", NULL,
+     0, FLB_FALSE, 0,
+    "A standard MIME type describing the format of the contents. This will be set"
```
Comment (PettitWesley) [inline plugins/out_s3/s3.c:1538]: A standard MIME type for the S3 object; this will be set as the Content-Type HTTP header

File: plugins/out_s3/s3.c (hunk 13 above)
Comment (zhonghui12) [reply plugins/out_s3/s3.c:493]: Sure.

File: plugins/out_s3/s3.c (hunk 14 above)
Comment (zhonghui12) [reply plugins/out_s3/s3.c:1012]: Yeah, but `void *` is the input type of `flb_gzip`. Although I think there is automatically conversion in C, I need to check if it will work.

File: plugins/out_s3/s3.c (hunk 15 above)
Comment (zhonghui12) [reply plugins/out_s3/s3.c:1532]: I see. Will change the word

File: plugins/out_s3/s3.c (hunk 14 above)
Comment (PettitWesley) [reply plugins/out_s3/s3.c:1012]: it should work. Even if it doesn't in this case you'd cast to void pointer when you make the call to flb_gzip_compress

**Review by PettitWesley — APPROVED** [review]

File: plugins/out_s3/s3.c [hunk 17]
```diff
@@ -50,6 +51,26 @@ static struct multipart_upload *get_upload(struct flb_s3 *ctx,
...
+};
+
+static struct flb_aws_header content_type_header = {
+    .key = "Content-Type",
+    .key_len = 12,
+    .val = "",
+    .val_len = 0,
+};
+
+struct flb_aws_header canned_acl_header = {
```
Comment (zhonghui12) [inline plugins/out_s3/s3.c:68]: I think I should add `static` here

Comment (PettitWesley) [reply plugins/out_s3/s3.c:68]: I agree

File: plugins/out_s3/s3.c [hunk 18]
```diff
@@ -1447,6 +1529,18 @@ static struct flb_config_map config_map[] = {
      0, FLB_FALSE, 0,
     "Predefined Canned ACL policy for S3 objects."
     },
+    {
+     FLB_CONFIG_MAP_STR, "compression", NULL,
+     0, FLB_FALSE, 0,
+    "Compression type for S3 objects. 'gzip' is currently the only supported value. "
+    "This will be set as the Content-Encoding HTTP header."
```
Comment (PettitWesley) [inline plugins/out_s3/s3.c:1536]: Just to be very clear, instead of saying "This will be set", say, "The Content-Encoding HTTP Header will be set to 'gzip'"
